torch/_functorch/eager_transforms.py

```diff
@@ -8,10 +8,16 @@
 import torch
 from functools import partial, wraps
 import contextlib
-from torch.utils._pytree import tree_flatten, tree_unflatten, tree_map, tree_map_only
+from torch.utils._pytree import (
+    tree_flatten,
+    tree_unflatten,
+    tree_map,
+    tree_map_only,
+    tree_map_,
+    treespec_pprint,
+)
```

Comment:
@XuehaiPan a couple of testing-related comments:
- Why do we have both TestCxxPytree and TestPytree? Ideally we would share code between the two so that we don't have to duplicate our tests
- Can we port some of the pytree tests from optree over to test_pytree.py? I want to make sure that PyTorch pytree and optree don't diverge and have correct semantics. I think the likely case is that optree is correct (because you folks based it off of JAX) and that PyTorch pytree may have different semantics because we wrote it in a hurryl

Reply:
> Why do we have both TestCxxPytree and TestPytree?

The Cxx pytree tests are almost identical. They can share most of the test cases. But they are different in some aspects.

1. The expected `treespec` object can be constructed from `TreeSpec` and `LeafSpec` for Python PyTree. But for `optree`, they can only obtained from `optree.tree_flatten(...)` or `optree.tree_structure(...)` because we do not want to expose too many implementation details (especially for dictionary ordering).
2. The definitions for `treespec_dumps` and `treespec_loads` are different in the two implementations. The serialization type for `torch.utils._pytree` is `str` and it is `bytes` for `torch.utils.pytree`.

> Can we port some of the pytree tests from optree over to test_pytree.py? 

Yes, we can. I think we can do this in a follow-up PR.

Reply:
> The Cxx pytree tests are almost identical. They can share most of the test cases. But they are different in some aspects.

We should figure out how to deduplicate the test cases before we ship this PR (ideally we would have done that in the previous PR). There are a number of options:
- a TestPyTreeGeneric test case which has the truly generic bits, and then TestPythonPyTree and TestCxxPyTree
- use the parametrized decorator on TestPyTree

> The definitions for treespec_dumps and treespec_loads are different in the two implementations. The serialization type for torch.utils._pytree is str and it is bytes for torch.utils.pytree.

It will be a problem if treespec_dumps/treespec_loads are different between Python and C++ pytree

>> Can we port some of the pytree tests from optree over to test_pytree.py?

> Yes, we can. I think we can do this in a follow-up PR.

My preference is that we first set up a PR with some testing (to confirm that PyTorch pytree has the desired semantics), and then we can add optree and see it also pass all of the test cases. If PyTorch pytree does not have the correct semantics, then we are going to need to figure out what to do about it (change PyTorch pytree semantics to be "correct"? add some special logic here in pytorch to make optree behave the same way?) But I can be convinced we go the other way.